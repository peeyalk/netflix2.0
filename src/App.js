import React, { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { Route, Routes } from 'react-router';
import { onAuthStateChanged } from 'firebase/auth';
import { auth } from './FirebaseApp';
import HomeScreen from './routes/HomeScreen';
import LoginScreen from './routes/Login';
import './App.css';
import {
  loginDispatchAction,
  logoutDispatchAction,
  selectAuthUser,
} from './redux/authUserReducer';
import { useSelector } from 'react-redux';
import ProfileScreen from './routes/ProfileScreen';
// import { store } from './app/store';

function App() {
  const dispatch = useDispatch();

  const currentAuthUser = useSelector(selectAuthUser);

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (authUser) => {
      if (authUser) {
        dispatch(loginDispatchAction(makeAuthUserFromFirebase(authUser)));
      } else {
        dispatch(logoutDispatchAction());
      }
    });

    return unsubscribe;
  }, [dispatch]);

  return (
    <div className='App'>
      <Routes>
        {!currentAuthUser ? (
          <>
            <Route path='/' element={<LoginScreen />} />

            {/* <Route path='/login' element={<LoginScreen />} /> */}
            {/* <Route path='/signin' element={<LoginScreen isSignInRoute={true} />} /> */}
          </>
        ) : (
          <>
            <Route path='/edit_profile' element={<ProfileScreen />} />
            <Route exact path='/' element={<HomeScreen />} />
          </>
        )}
      </Routes>
    </div>
  );
}

export const makeAuthUserFromFirebase = (authUser) => {
  return {
    uid: authUser.uid,
    email: authUser.email,
    refreshToken: authUser.stsTokenManager.refreshToken,
    accessToken: authUser.stsTokenManager.accessToken,
    expirationTime: authUser.stsTokenManager.expirationTime,
    createdAt: authUser.createdAt,
    lastLoginAt: authUser.lastLoginAt,
  };
};

export default App;
