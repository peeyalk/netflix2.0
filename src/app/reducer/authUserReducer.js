import { createSlice } from '@reduxjs/toolkit';

const authUserReducer = createSlice({
  name: 'authUser',
  initialState: {
    authUser: null,
  },
  reducers: {
    loginDispatchAction: (state, action) => {
      state.authUser = action.payload;
    },
    logoutDispatchAction: (state) => {
      state.authUser = null;
    },
  },
});

export const { loginDispatchAction, logoutDispatchAction } = authUserReducer.actions;

/**
 * To get this state
 * call this function with useSelector hook ie:
 * const currentAuthUser = useSelector(selectAuthUser);
 */
export const selectAuthUser = (state) => state.authUserReducer.authUser;

export default authUserReducer.reducer;
