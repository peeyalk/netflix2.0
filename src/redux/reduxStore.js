import { configureStore } from '@reduxjs/toolkit';
import authUserReducer from './authUserReducer';

export const reduxStore = configureStore({
  reducer: {
    authUserReducer: authUserReducer,
  },
});
