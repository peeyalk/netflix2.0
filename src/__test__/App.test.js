import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import App, { makeAuthUserFromFirebase } from '../App';
import { BrowserRouter } from 'react-router-dom';
import { unmountComponentAtNode } from 'react-dom';
import { reduxStore } from '../redux/reduxStore';
import { loginDispatchAction } from '../redux/authUserReducer';
import { act } from 'react-dom/test-utils';
import AuthUserData from './authUser.json';

let container = null;
// setup a DOM element as a render target
beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});
// Cleanup after each test
afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

test('App renders without crashing', () => {
  render(
    <BrowserRouter>
      <React.StrictMode>
        <Provider store={reduxStore}>
          <App />
        </Provider>
      </React.StrictMode>
    </BrowserRouter>,
    container
  );
});

// Unauthenticated user should see the login screen
test('Unauthenticated user should see the login screen', () => {
  const { asFragment, getByText } = render(
    <BrowserRouter>
      <React.StrictMode>
        <Provider store={reduxStore}>
          <App />
        </Provider>
      </React.StrictMode>
    </BrowserRouter>,
    container
  );

  expect(
    getByText(/Unlimited movies, TV shows, and anime./i)
  ).toBeInTheDocument();
  expect(getByText(/Watch anywhere. Cancel anytime./i)).toBeInTheDocument();
  expect(
    getByText(/Enter your email to create or restart your membership./i)
  ).toBeInTheDocument();
  expect(getByText(/Get Started/i)).toBeInTheDocument();
  expect(asFragment()).toMatchSnapshot();
});

// Authenticated user should see the home page
test('Authenticated user should see the home page', () => {
  // mock authenticated user
  act(() => {
    /* fire events that update state */
    reduxStore.dispatch(
      loginDispatchAction(makeAuthUserFromFirebase(AuthUserData))
    );
  });

  const { asFragment, getByText } = render(
    <BrowserRouter>
      <React.StrictMode>
        <Provider store={reduxStore}>
          <App />
        </Provider>
      </React.StrictMode>
    </BrowserRouter>,
    container
  );

  expect(getByText(/#1 in Movies Today/i)).toBeInTheDocument();
  expect(asFragment()).toMatchSnapshot();
});

// Redux Store should have the same Authenticated user data
test('Redux Store should have the same Authenticated user data, () => {', () => {
  reduxStore.dispatch(
    loginDispatchAction(makeAuthUserFromFirebase(AuthUserData))
  );
  expect(reduxStore.getState().authUserReducer.authUser).toEqual(
    makeAuthUserFromFirebase(AuthUserData)
  );
});
