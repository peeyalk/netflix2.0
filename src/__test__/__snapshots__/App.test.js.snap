// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authenticated user should see the home page 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="homeScreen"
    >
      <div
        class="nav false"
      >
        <div
          class="nav__contents"
        >
          <img
            alt="logo"
            class="nav__logo"
            src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Logonetflix.png"
          />
          <img
            alt="avatar"
            class="nav__avatar"
            src="https://avatars.githubusercontent.com/u/17343278?s=40&v=4"
          />
        </div>
      </div>
      <header
        class="banner"
        style="background-size: cover;"
      >
        <div
          class="banner__contents"
        >
          <h1
            class="banner__title"
          />
          <h2
            class="banner__supplemental-message"
          >
            <img
              src="no1MovieToday.svg"
              style="margin-right: 15px;"
            />
            #1 in Movies Today
          </h2>
          <h1
            class="banner__discription"
          />
          <div
            class="banner__buttons"
          >
            <button
              class="banner__button"
            >
              Play
            </button>
            <button
              class="banner__button"
            >
              My List
            </button>
          </div>
        </div>
        <div
          class="banner--fadeButtom"
        />
      </header>
      <div
        class="row"
      >
        <h2>
          NETFLIX ORIGINAL
        </h2>
        <div
          class="row__posters"
        />
      </div>
      <div
        class="row"
      >
        <h2>
          Trending Now
        </h2>
        <div
          class="row__posters"
        />
      </div>
      <div
        class="row"
      >
        <h2>
          Top Rated
        </h2>
        <div
          class="row__posters"
        />
      </div>
      <div
        style="height: 40px;"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Unauthenticated user should see the login screen 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="loginScreen"
    >
      <div
        class="loginScreen__background"
      >
        <img
          alt="netflix2.0"
          class="loginScreen__logo"
          src="netflix_logo.png"
        />
        <button
          class="loginScreen__button"
        >
          Sign In
        </button>
        <div
          class="loginScreen__gradient"
        />
      </div>
      <div
        class="loginScreen__body"
      >
        <h1>
          Unlimited movies, TV shows, and anime.
        </h1>
        <h2>
          Watch anywhere. Cancel anytime.
        </h2>
        <h3>
          Ready to watch? Enter your email to create or restart your membership.
        </h3>
        <div
          class="loginScreen__input"
        >
          <form>
            <input
              placeholder="Email Address"
              type="email"
            />
            <button
              class="loginScreen__getStarted_btn"
            >
              Get Started
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
